### PR TITLE
[oneCCL] enable bf16 comm.

### PR DESCRIPTION
**Pull request by marvin-Yu**
Enable communication using XFT_ONECCL_BF16 for bf16 data type.

### CODE REVIEW

File: src/common/bfloat16.h [hunk 1]
```diff
@@ -78,7 +78,9 @@ class bfloat16_t {
     }
 
     static void cvt_float_to_bfloat16(const float *src, bfloat16_t *dst, int size);
+    static void batch_cvt_float_to_bfloat16(const float *src, bfloat16_t *dst, int size);
```
Comment (pujiang2018) [inline src/common/bfloat16.h:81]: @changqi1 Do you accept adding it in bfloat16.h?

File: src/utils/messenger.h [hunk 2]
```diff
@@ -39,6 +42,12 @@ class Messenger {
             return;
         }
 
+        bf16_enable = (getenv("XFT_ONECCL_BF16") ? atoi(getenv("XFT_ONECCL_BF16")) : 0);
+        if (bf16_enable) {
+            printf("got 'XFT_ONECCL_BF16=%d', enable BF16 dtype comm.\n", bf16_enable);
+            buf_bf16 = new bfloat16_t[MAX_BF16_BUFFER];
```
Comment (pujiang2018) [inline src/utils/messenger.h:48]: suggest using aligned_alloc or malloc.
as new will initialize all the data to 0, which is not needed here.

File: src/utils/messenger.h [hunk 3]
```diff
@@ -110,7 +123,20 @@ class Messenger {
...
     void reduceAdd(T *sendBuf, T *recvBuf, size_t count) {
+        if constexpr (!std::is_same_v<T, float>) {
+            //todo(marvin): Consideration for additional optimization scenarios
+            printf("Currently unsupport about data types other than float.\n");
+            exit(-1);
+        }
+
         TimeLine t("Messenger.reduceAdd");
+        if (bf16_enable) {
+            bfloat16_t::batch_cvt_float_to_bfloat16(sendBuf, buf_bf16, count);
```
Comment (pujiang2018) [inline src/utils/messenger.h:134]: what if count > MAX_BF16_BUFFER?

File: src/utils/messenger.h [hunk 4]
```diff
@@ -110,7 +123,20 @@ class Messenger {
...
+            ccl::allreduce(
+            buf_bf16, buf_bf16, count, ccl::datatype::bfloat16, ccl::reduction::sum, *pcomm).wait();
```
Comment (pujiang2018) [inline src/utils/messenger.h:136]: Here the code looks like not formatted.

File: src/utils/messenger.h [hunk 5]
```diff
@@ -175,7 +201,8 @@ class Messenger {
     int size;
     int rank;
     bool local_ranks_flag;
-
+    bfloat16_t* buf_bf16 = nullptr;
```
Comment (pujiang2018) [inline src/utils/messenger.h:204]: let's initialize it in the constructor to make sure the same style.

File: src/common/bfloat16.h (hunk 1 above)
Comment (changqi1) [reply src/common/bfloat16.h:81]: @pujiang2018  @marvin-Yu Sure. It is great to add batch cvt. But the number 1024 is not make sense.

File: src/common/bfloat16.h [hunk 6]
```diff
@@ -128,6 +130,18 @@ inline void bfloat16_t::cvt_float_to_bfloat16(const float *src, bfloat16_t *dst,
     }
 }
 
+inline void bfloat16_t::batch_cvt_float_to_bfloat16(const float *src, bfloat16_t *dst, int count){
+    constexpr int sizePerSplit = 1024;
```
Comment (changqi1) [inline src/common/bfloat16.h:134]: Maybe use template to define blockSize or use totalSize/totalSystemThreads to define the blockSize.

File: src/utils/messenger.h (hunk 2 above)
Comment (changqi1) [reply src/utils/messenger.h:48]: Please add this macro XFT_ONECCL_BF16 into wiki.

Comment (billishyahao) [inline src/utils/messenger.h:48]: Init buf_bf16 with max length may cause performance communication issue, please double confirm this.

**Pull request closed without merging**